### PR TITLE
Update Babel roadmap

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
This was an internal roadmap written by me and other team members for a possible sponsorship. We can make it public, to show the future direction of the project (or at least, its how the current core team would like to evolve the project).

I didn't delete the old roadmap (from 3 years ago), I just commented it out. About half of the things it contained have been done; the others seem more like ideas that are hard to implement with the Babel architecture.

### CODE REVIEW

**Review by hzoo — APPROVED** [review]
Oh I guess we can add this back to the sidebar somewhere?

File: docs/roadmap.md [hunk 1]
```diff
@@ -4,6 +4,149 @@ title: Babel Roadmap
 sidebar_label: Roadmap
 ---
 
+This document outlines some of the improvements our team members would like to work on during this year.
+
+This is far from being a complete list of all the new features or important changes that we'll bring to Babel, but it's a good summary if you are interested in the general direction that the project is moving to. We might actually not do every listed point, or we might delay some of them to the next year. Some of them have clear starting and ending points, while others need more research or [RFCs](https://github.com/babel/rfcs).
```
Comment (existentialism) [inline docs/roadmap.md:9]:
```suggestion
This is far from being a complete list of all the new features or important changes that we'll bring to Babel, but it's a good summary if you are interested in the general direction that the project is toward. We might actually not do every listed point, or we might delay some of them to the next year. Some of them have clear starting and ending points, while others need more research or [RFCs](https://github.com/babel/rfcs).
```

File: docs/roadmap.md [hunk 2]
```diff
@@ -4,6 +4,149 @@ title: Babel Roadmap
...
+
+If your company is interested in some of these and would like to directly sponsor their implementation please [reach out](mail:team@babeljs.io)!
+
+## Babel 2021 Roadmap
+
+### Babel 8
+
+We have been talking about the Babel 8 release for more than one year (actually, we initially scheduled it for about one year ago)! However, we are now closer then ever to actually release it.
```
Comment (existentialism) [inline docs/roadmap.md:17]:
```suggestion
We have been talking about the Babel 8 release for more than one year (actually, we initially scheduled it about one year ago)! However, we are now closer then ever to its release!
```

File: docs/roadmap.md [hunk 3]
```diff
@@ -4,6 +4,149 @@ title: Babel Roadmap
...
+
+Most of the remaining tasks are in the [tracking issue](https://github.com/babel/babel/issues/10746), but there are still a few blockers:
+- We want to drop support for [Node.js 10](https://github.com/nodejs/Release), which stops being maintained on 2021-04-30;
+- We would like to release Babel as a pure ESM package. We are now in the process of converting our sources to be compatible with Node.js' ESM implementation, and while doing it we are examining how we can make it easier for people that currently use Babel to compile ESM to CJS;
+- We are trying to align our TypeScript AST with the `typescript-eslint` project. Our ASTs are _almost_ identical, but we need to introduce some small breaking changes to fully align;
+- Our release infrastructure doesn't support yet pre-releases, or using multiple "main" branches (one for Babel 8 and one for Babel 7);
+- We haven't figured out yet a policy for Babel 7 maintainance.
```
Comment (existentialism) [inline docs/roadmap.md:24]:
```suggestion
- We haven't figured out yet a policy for Babel 7 maintenance.
```

File: docs/roadmap.md [hunk 4]
```diff
@@ -4,6 +4,149 @@ title: Babel Roadmap
...
+
+We support all the Stage 2 proposals except for:
+- The new iteration of the decorators proposal (we need to implement both parsing and transform);
+- Transform for the Module Blocks proposal (we implemented parsing in Babel 7.13.0).
+
+We will implement support for decorators, and investigate if and how we can implement transform for module blocks.
+
+While we don't support many Stage 1 proposals, there have been recent updates to the pipeline operator and to do expressions: since we already support those proposals and the community is quite excited about them, we will update our implementations.
+
+There are also other proposal, such as pattern matching, that we didn't implement yet because their champions expect to do significant changes to the syntax and semantics. However, we are closely following their development to implement them in Babel as soon as they stabilize a bit.
```
Comment (existentialism) [inline docs/roadmap.md:38]:
```suggestion
There are also other proposals, such as pattern matching, that we have not yet implement because their champions expect significant changes to the syntax and semantics. However, we are closely following their development, and will implement them in Babel as soon as they stabilize a bit.
```

File: docs/roadmap.md [hunk 5]
```diff
@@ -4,6 +4,149 @@ title: Babel Roadmap
...
+### Move `@babel/preset-env` into `@babel/core`
+
+A minimal Babel transforming setup requires at least three packages:
+- `@babel/core`
+- `@babel/preset-env`
+- a Babel "runner" (`@babel/cli`, `babel-loader`, `@rollup/plugin-babel`, etc)
+
+Moving `@babel/preset-env` directly into `@babel/core` has two big advantages:
+- It will be easier to configure Babel in simple projects, you would only need to enable a `compileJS: true` option in `babel.config.json` (or it could even be the default in the future -- it can't be default as `@babel/eslint-parser` does not compile the source)
+- It will make sure that the plugin versions are in sync with the `@babel/core` version, avoiding most of the bugs casued by accidentally incompatible packages versions
```
Comment (existentialism) [inline docs/roadmap.md:49]:
```suggestion
- It will make sure that the plugin versions are in sync with the `@babel/core` version, avoiding most of the bugs caused by mismatched/incompatible packages versions
```

File: docs/roadmap.md [hunk 6]
```diff
@@ -4,6 +4,149 @@ title: Babel Roadmap
...
+- When we'll start moving to ESM, it will be hard to resolve and load plugin synchronously in `transformSync`. This prevents it from being a problem.
```
Comment (existentialism) [inline docs/roadmap.md:50]:
```suggestion
- When we move to ESM, it will be hard to resolve and load plugin synchronously in `transformSync`. This prevents it from being a problem.
```

File: docs/roadmap.md [hunk 7]
```diff
@@ -4,6 +4,149 @@ title: Babel Roadmap
...
+
+There is already [a RFC](https://github.com/babel/rfcs/pull/10) to move _plugins_ for stable ECMAScript features in `@babel/core`, which is the first step in this direction.
+
+With our current `@babel/preset-env` architecture, we would need to specially handle official plugins to automatically enable or disable them based on `targets`.
+However, this has two drawbacks:
+- Compatibility data for a specific plugin is completely separated from the plugin implementation (it's not even a dependency, more something like an internal implicit peer dependency: plugin -> @babel/core -> @babel/compat-data);
+- Official plugins would get a special treatment from `@babel/core`, while we want to make sure that third-party plugins have the same capabilities as official plugins.
```
Comment (existentialism) [inline docs/roadmap.md:57]:
```suggestion
- Official plugins would get special treatment from `@babel/core`, while we want to make sure that third-party plugins have the same capabilities as official plugins.
```

File: docs/roadmap.md [hunk 8]
```diff
@@ -4,6 +4,149 @@ title: Babel Roadmap
...
+
+### Continue developing the `babel-polyfills` project
+
+We have already decided to remove the old `core-js@2` support from `@babel/preset-env` in Babel 8, and we also want to stop promoting a specific third-party polyfill giving our users the impression that it's part of Babel itself.
```
Comment (existentialism) [inline docs/roadmap.md:61]:
```suggestion
We have already decided to remove the old `core-js@2` support from `@babel/preset-env` in Babel 8. We also want to stop promoting a specific third-party polyfill, which might give our users the impression that it's part of Babel itself.
```

File: docs/roadmap.md [hunk 9]
```diff
@@ -4,6 +4,149 @@ title: Babel Roadmap
...
+
+This might happen in two different ways:
+- We just remove `core-js@3` from `@babel/preset-env` in Babel 8, encouraging users to migrate to `babel-plugin-polyfill-corejs3` (which is what `@babel/preset-env` internally uses starting from version 7.10.0)
+- We can keep `core-js@3` support in `@babel/preset-env`, but not migrate it to `@babel/core` when we'll move the transform plugins.
+
+Whatever path we take, we would like to offer at least one alternative to our users when they'll need to update the `core-js` integration in their configuration. `core-js` is a really good polyfill that ensures the highest possible spec compliancy, but different users might prefer different trade-offs.
```
Comment (existentialism) [inline docs/roadmap.md:67]:
```suggestion
Whichever path we take, we would like to offer at least one alternative to our users when they'll need to update the `core-js` integration in their configuration. `core-js` is a really good polyfill that ensures the highest possible spec compliancy, but users may prefer different trade-offs.
```

File: docs/roadmap.md [hunk 10]
```diff
@@ -4,6 +4,149 @@ title: Babel Roadmap
...
+
+([Nicolò](https://github.com/nicolo-ribaudo)) is working with [@ljharb](https://github.com/ljharb) to make sure that the [`@es-shims` project](https://github.com/es-shims/) supports at least all the ES2015+ features (we actually aim for ES5+), so that Babel users are free to choose to between at least two alternatives.
```
Comment (existentialism) [inline docs/roadmap.md:69]:
```suggestion
([Nicolò](https://github.com/nicolo-ribaudo)) is working with [@ljharb](https://github.com/ljharb) to make sure that the [`@es-shims` project](https://github.com/es-shims/) supports at least all the ES2015+ features (we actually aim for ES5+), so that Babel users are free to choose to between at least two options.
```

File: docs/roadmap.md [hunk 11]
```diff
@@ -4,6 +4,149 @@ title: Babel Roadmap
...
+### Expand `targets` usage for granular transforms
+
+Since the beginning, `@babel/preset-env` has used the `targets` option to automatically enable or disable transform plugins.
+
+However, there isn't a 1-to-1 mapping between Babel plugins and features implemented in browsers.
+
+For example, we have a single plugin for the different class fields types (public and private, static and instance), but browsers have varying compatibility matrices:
+
+- Firefox 73 and Safari 14 support only public instance fields
+- Firefox 75+ supports pulic instance and static fields
```
Comment (existentialism) [inline docs/roadmap.md:82]:
```suggestion
- Firefox 75+ supports public instance and static fields
```

File: docs/roadmap.md [hunk 12]
```diff
@@ -4,6 +4,149 @@ title: Babel Roadmap
...
+- Chrome 79+ supports public and private fields, but doesn't support private fields in some optional chaining expressions
+- Chrome 84+ fully supports private fields, and also private methods
+- Safari TP 121 fully supports private fields (even with `?.`), but it doesn't support private methods
+
+Creating a plugin for each feature is sub-optimal. For example, we can convert private methods to private fields and then, if needed, convert them to older syntax. However, we can generate a better optimized output by directly converting private methods down to older syntax without the intermediate step, if we know that it needs to be transpiled down.
```
Comment (existentialism) [inline docs/roadmap.md:87]:
```suggestion
Creating a plugin for each feature is sub-optimal. For example, we can convert private methods to private fields and then, if needed, convert them to older syntax. However, we can generate better/optimized output by directly converting private methods down to older syntax without the intermediate step if we know that it needs to be transpiled down.
```

File: docs/roadmap.md [hunk 13]
```diff
@@ -4,6 +4,149 @@ title: Babel Roadmap
...
+
+Since starting from Babel 7.13.0 we can read `targets` option directly inside a plugin, we can modify our plugins to automatically perform _partial_ compilation of a given ECMAScript feature, where this would give an advantage in the output size and runtime performance.
```
Comment (existentialism) [inline docs/roadmap.md:89]:
```suggestion
Since Babel 7.13.0, we can read the `targets` option directly inside a plugin, we can modify our plugins to automatically perform _partial_ compilation of a given ECMAScript feature, where this would give an advantage in the output size and runtime performance.
```

File: docs/roadmap.md (hunk 3 above)
Comment (nicolo-ribaudo) [reply docs/roadmap.md:24]: 50% of the times I search on google how to spell it, the other 50% I spell it wrong :joy:

File: docs/roadmap.md [hunk 14]
```diff
@@ -4,6 +4,149 @@ title: Babel Roadmap
...
+
+**Prior Art**
+
+This approach isn't completely new. Thanks to a collaboration with [@_developit](https://twitter.com/_developit), in Babel 7.9.0 we introduced a new `bugfixes: true` option for `@babel/preset-env`. When this option is enabled, and when using `esmodules: true` as the compilation targets, we only partially compile [some features](https://github.com/babel/preset-modules#features-supported). This what made us initially think about this possibility, but the current partial transforms are less useful when using more modern targets (for example, `defaults, not ie 11`).
```
Comment (existentialism) [inline docs/roadmap.md:94]:
```suggestion
This approach isn't completely new. Thanks to a collaboration with [@_developit](https://twitter.com/_developit), in Babel 7.9.0 we introduced a new `bugfixes: true` option for `@babel/preset-env`. When this option is enabled, and when using `esmodules: true` as the compilation target, we only partially compile [some features](https://github.com/babel/preset-modules#features-supported). This what made us initially think about this possibility, but the current partial transforms are less useful when using more modern targets (for example, `defaults, not ie 11`).
```

File: docs/roadmap.md [hunk 15]
```diff
@@ -4,6 +4,150 @@ title: Babel Roadmap
...
+
+We also already use the `targets` option to decide whether we can use `Object.assign` when compiling object spread or not.
+
+**Action Points**
+
+This goal can be divided in two main big tasks, that can be carried on in parallel:
```
Comment (existentialism) [inline docs/roadmap.md:100]:
```suggestion
This goal can be split into two big tasks that can be done in parallel:
```

File: docs/roadmap.md [hunk 16]
```diff
@@ -4,6 +4,150 @@ title: Babel Roadmap
...
+- Manually checking which features we compile to "non-obvious" output, which is usually caused by edge cases that many developers don't care about.
+- Ask for feedback from the community, since developers can test which assumptions work and which don't on their applications.
+
+### Overhaul the Babel REPL
+
+The Babel REPL is a conveninent playground to learn how Babel transpiles source code.
+
+Current limits:
+
+- The REPL does not support `assumptions` config. Although we have delicated per-assumption basis mini REPL on https://babel.dev/assumptions, currently we can not show how these `assumptions` may work toghether
```
Comment (existentialism) [inline docs/roadmap.md:124]:
```suggestion
- The REPL does not support `assumptions` config. Although we have a dedicated per-assumption basis mini REPL on https://babel.dev/assumptions, currently we can not show how these `assumptions` might work together
```

File: docs/roadmap.md [hunk 17]
```diff
@@ -4,6 +4,150 @@ title: Babel Roadmap
...
+https://github.com/babel/website/issues/1292, https://github.com/babel/website/issues/2224, https://github.com/babel/website/pull/1970
+
+Features good to have:
+
+- AST Explorer (integrate with existing one)
+- stderr with complete stack trace as error log
+- stdout as output
+- Change Babel version from UI
+
+At least 15% of open issues in babel-website is related to REPL: https://github.com/babel/website/issues?q=is%3Aissue+is%3Aopen+label%3Arepl
```
Comment (existentialism) [inline docs/roadmap.md:135]:
```suggestion
At least 15% of open issues in babel-website are related to the REPL: https://github.com/babel/website/issues?q=is%3Aissue+is%3Aopen+label%3Arepl
```

**Review by existentialism — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
:rocket:
